### PR TITLE
Database split part III: Use two different postgresql databases.

### PR DESCRIPTION
**Pull request by noliveleger**
This PR is based on #2186 and #2185 but really uses two distincts databases.
One for `kpi` and one for `kc`. 
A script has been created to copy data from `kc` to `kpi`. 

Needs to be merged with:
- `kobo-docker` https://github.com/kobotoolbox/kobo-docker/pull/230
- `kobo-install` https://github.com/kobotoolbox/kobo-install/pull/15
- `kc` https://github.com/kobotoolbox/kobocat/pull/531

It now uses redis to store session instead of PostgreSQL

### CODE REVIEW

File: dependencies/pip/dev_requirements.txt [hunk 1]
```diff
@@ -134,6 +135,6 @@ werkzeug==0.14.1
 whitenoise==3.3.1
 whoosh==2.7.4
 xlrd==1.1.0
-xlsxwriter==1.0.4
+xlsxwriter==1.1.2
```
Comment (jnm) [inline dependencies/pip/dev_requirements.txt]: :clap: :clap:

File: kobo/settings/base.py [hunk 2]
```diff
@@ -695,3 +704,7 @@ def __init__(self, *args, **kwargs):
 MONGO_CONNECTION = MongoClient(
     MONGO_CONNECTION_URL, j=True, tz_aware=True, connect=False)
 MONGO_DB = MONGO_CONNECTION[MONGO_DATABASE['NAME']]
+
+
+SESSION_ENGINE = "redis_sessions.session"
+SESSION_REDIS = RedisHelper.config(default="redis://redis_cache:6380/2")
```
Comment (jnm) [inline kobo/settings/base.py]: Make sure that `redis_cache` has persistent storage

File: kpi/deployment_backends/kc_access/shadow_models.py [hunk 3]
```diff
@@ -202,6 +204,65 @@ def save(self, *args, **kwargs):
...
+
+        kc_auth_user.save()
+
+
+class KCToken(ShadowModel):
+
+    key = models.CharField(_("Key"), max_length=40, primary_key=True)
+    user = models.OneToOneField(getattr(settings, 'AUTH_USER_MODEL', 'auth.User'),
+                                related_name='auth_token',
+                                on_delete=models.CASCADE, verbose_name=_("User"))
```
Comment (jnm) [inline kpi/deployment_backends/kc_access/shadow_models.py]: Note to self: look at a reference to see which direction the `CASCADE` goes :heavy_check_mark: Looks good :+1:

File: kpi/deployment_backends/kc_access/shadow_models.py [hunk 4]
```diff
@@ -202,6 +204,65 @@ def save(self, *args, **kwargs):
...
+    created = models.DateTimeField(_("Created"), auto_now_add=True)
+
+    class Meta(ShadowModel.Meta):
+        db_table = "authtoken_token"
+
+    @classmethod
+    def sync(cls, auth_token):
+        try:
+            kc_auth_token = cls.objects.get(pk=auth_token.pk)
```
Comment (jnm) [inline kpi/deployment_backends/kc_access/shadow_models.py]: Thought:

1. Make a new token (`123`) on KPI. The PK is the token value (`123`);
1. Sync token to KC, saving token `123` in KC's database;
1. Delete KPI token `123`;
1. Somehow, the token delete signal fails, and the KC token `123` remains in the database(!)
1. Make a new KPI token (`456`);
1. Call `sync()` here: KC token `456` does not exist, so we make a new one. However, KC token `123` still exists and is linked to the user. An `IntegrityError` is raised.

Could we look up the token by user instead?

File: kpi/utils/redis_helper.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    """
+
+    @staticmethod
+    def config(default=None):
+        """
+        :return: dict
+        """
+
+        try:
```
Comment (jnm) [inline kpi/utils/redis_helper.py]: We can omit the `try` (see next comment)

File: kpi/utils/redis_helper.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            redis_connection_url = os.getenv("REDIS_SESSION_URL", default)
+            match = re.match(r"redis://(:(?P<password>[^@]*)@)?(?P<host>[^:]+):(?P<port>\d+)(/(?P<index>\d+))?",
+                             redis_connection_url)
+            if not match:
+                raise Exception()
```
Comment (jnm) [inline kpi/utils/redis_helper.py]: Yes, let's raise an exception here (and not catch it). Let's use `ImproperlyConfigured` and the text already included below (`"Could not parse Redis session URL. Please verify 'REDIS_SESSION_URL' value"`)

File: kobo/settings/base.py (hunk 2 above)
Comment (noliveleger) [reply kobo/settings/base.py]: It does. 
From `docker-compose.backend.template.yml` in `kobo-docker` related branch. 

```
  redis_cache:
    image: redis:3.2
    # Map our "cache" Redis config into the container.
    volumes:
      - ./.vols/redis_cache_data/:/data/
```

File: kpi/deployment_backends/kc_access/shadow_models.py (hunk 4 above)
Comment (noliveleger) [reply kpi/deployment_backends/kc_access/shadow_models.py]: It did the trick. 
I changed it.

**Pull request merged**